rust-runtime/aws-smithy-http-server/src/request.rs

```diff
@@ -0,0 +1,111 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+//! Parse HTTP request trait
+
+use aws_smithy_http::body::SdkBody;
+use bytes::Bytes;
+
+/// `ParseHttpRequest` is a generic trait for parsing structured data from HTTP requests.
+///
+/// It is designed to be flexible, because `Input` is unconstrained, it can be used to support
+/// event streams, regular request-response style operations, as well as any other HTTP-based
+/// protocol that we manage to come up with.
+///
+/// It also enables this critical and core trait to avoid being async, and it makes code that uses
+/// the trait easier to test.
+pub trait ParseHttpRequest {
+    /// Input type of the HttpRequest
+    ///
+    /// For request/response style operations, this is typically something like:
+    /// `Result<OperationInput, OperationInputError>`
+    ///
+    /// For streaming operations, this is something like:
+    /// `Result<EventStream<OperationEvent>>, OperationStreamingError>`
+    type Input;
```

Comment:
Similar idea here: why doesn't the `Input` type directly `impl` `ParseHttpRequest`?

Reply:
Yep, you are right, will fix it.

Reply:
doing it this way has some advantages, namely, that you can give the parser state that the input doesn't have. Eg:

```rust
struct MyInputParser {
  a: String
  bunch: bool,
  of: u64
  parsing_settings: Bytes
}

impl ParseHttpRequest for MyInputParser {
  ...
}
```
It also avoids the orphan trait rule if you ever want to define parsers for something that's actually in an external crate.